File: test/hardware/unit/test_memory_subsystem.py
import cocotb
from cocotb.triggers import RisingEdge, Timer
from cocotb.clock import Clock
from pathlib import Path
from test.driver import run_hardware_test
from test.env import get_rtl_dir


@cocotb.test()
async def test_memory_subsystem_basic(dut):
    """Test memory subsystem with simple read/write operations."""
    
    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())
    
    # Reset
    dut.rst_n.value = 0
    dut.icache_mem_req.value = 0
    dut.dcache_mem_req.value = 0
    dut.dcache_mem_we.value = 0
    await Timer(20, units="ns")
    await RisingEdge(dut.clk)
    dut.rst_n.value = 1
    
    # Wait for memory initialization to complete
    for _ in range(10):
        await RisingEdge(dut.clk)
    
    # Test 1: I-Cache read
    dut.icache_mem_addr.value = 0x0
    dut.icache_mem_req.value = 1
    
    # Wait for ready signal (simulated latency)
    for i in range(10):
        await RisingEdge(dut.clk)
        if dut.icache_mem_ready.value == 1:
            break
    
    assert dut.icache_mem_ready.value == 1, "I-Cache read should complete"
    try:
        val = int(dut.icache_mem_rdata.value)
        dut._log.info(f"I-Cache read: addr=0x0, data=0x{val:08x}")
    except ValueError:
        dut._log.info(f"I-Cache read: addr=0x0, data={dut.icache_mem_rdata.value}")
    
    dut.icache_mem_req.value = 0
    await RisingEdge(dut.clk)
    
    # Test 2: D-Cache write
    dut.dcache_mem_addr.value = 0x1000
    dut.dcache_mem_wdata.value = 0xDEADBEEF
    dut.dcache_mem_be.value = 0b1111
    dut.dcache_mem_we.value = 1
    dut.dcache_mem_req.value = 1
    
    # Wait for ready
    for i in range(10):
        await RisingEdge(dut.clk)
        if dut.dcache_mem_ready.value == 1:
            break
    
    assert dut.dcache_mem_ready.value == 1, "D-Cache write should complete"
    
    dut.dcache_mem_req.value = 0
    dut.dcache_mem_we.value = 0
    await RisingEdge(dut.clk)
    
    # Test 3: D-Cache read back
    dut.dcache_mem_addr.value = 0x1000
    dut.dcache_mem_req.value = 1
    dut.dcache_mem_we.value = 0
    
    for i in range(10):
        await RisingEdge(dut.clk)
        if dut.dcache_mem_ready.value == 1:
            break
    
    assert dut.dcache_mem_ready.value == 1, "D-Cache read should complete"
    assert dut.dcache_mem_rdata.value == 0xDEADBEEF, f"Read data mismatch: got 0x{dut.dcache_mem_rdata.value:08x}"
    
    dut._log.info("Memory Subsystem test passed")


def test_memory_subsystem():
    rtl_dir = get_rtl_dir()
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="memory_subsystem",
        verilog_sources=[
            rtl_dir / "system" / "memory_subsystem.v",
            rtl_dir / "memory" / "main_memory.v"
        ]
    )

File: test/hardware/unit/test_memory_subsystem.cpp
#include "tb_base.h"
#include "Vmemory_subsystem.h"
#include <string>

class MemorySubsystemTestbench : public ClockedTestbench<Vmemory_subsystem> {
public:
    MemorySubsystemTestbench() : ClockedTestbench<Vmemory_subsystem>(200, true, "memory_subsystem_trace.vcd") {
        // Initialize inputs
        dut->icache_mem_req = 0;
        dut->icache_mem_addr = 0;
        dut->dcache_mem_req = 0;
        dut->dcache_mem_addr = 0;
        dut->dcache_mem_wdata = 0;
        dut->dcache_mem_be = 0;
        dut->dcache_mem_we = 0;
    }
    
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }
    
    void reset() {
        dut->rst_n = 0;
        tick();
        dut->rst_n = 1;
        
        // Wait for memory initialization
        for (int i = 0; i < 10; i++) {
            tick();
        }
        TB_LOG("Reset complete");
    }
    
    void test_icache_read() {
        TB_LOG("Test: I-Cache read");
        
        dut->icache_mem_addr = 0x0;
        dut->icache_mem_req = 1;
        
        // Wait for ready signal
        bool ready = false;
        for (int i = 0; i < 50; i++) {
            tick();
            if (dut->icache_mem_ready == 1) {
                ready = true;
                break;
            }
        }
        
        TB_ASSERT_EQ(ready, true, "I-Cache read should complete");
        printf("[DEBUG] I-Cache read: addr=0x0, data=0x%08x\n", dut->icache_mem_rdata);
        
        dut->icache_mem_req = 0;
        tick();
    }
    
    void test_dcache_write() {
        TB_LOG("Test: D-Cache write");
        
        dut->dcache_mem_addr = 0x1000;
        dut->dcache_mem_wdata = 0xDEADBEEF;
        dut->dcache_mem_be = 0b1111;
        dut->dcache_mem_we = 1;
        dut->dcache_mem_req = 1;
        
        // Wait for ready
        bool ready = false;
        for (int i = 0; i < 50; i++) {
            tick();
            if (dut->dcache_mem_ready == 1) {
                ready = true;
                break;
            }
        }
        
        TB_ASSERT_EQ(ready, true, "D-Cache write should complete");
        
        dut->dcache_mem_req = 0;
        dut->dcache_mem_we = 0;
        tick();
    }
    
    void test_dcache_read() {
        TB_LOG("Test: D-Cache read back");
        
        dut->dcache_mem_addr = 0x1000;
        dut->dcache_mem_req = 1;
        dut->dcache_mem_we = 0;
        
        // Wait for ready
        bool ready = false;
        for (int i = 0; i < 50; i++) {
            tick();
            if (dut->dcache_mem_ready == 1) {
                ready = true;
                break;
            }
        }
        
        TB_ASSERT_EQ(ready, true, "D-Cache read should complete");
        TB_ASSERT_EQ(dut->dcache_mem_rdata, 0xDEADBEEF, "Read data matches write");
        
        dut->dcache_mem_req = 0;
        tick();
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    try {
        MemorySubsystemTestbench tb;
        
        tb.reset();
        tb.test_icache_read();
        tb.test_dcache_write();
        tb.test_dcache_read();
        
        TB_LOG("All Memory Subsystem tests PASSED!");
        return 0;
        
    } catch (const std::exception& e) {
        TB_ERROR(e.what());
        return 1;
    }
}
